{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Exploring the Model Control Plane Backend: A Practical Guide\n",
    "Welcome to our tour into the Model Control Plane, where we'll focus on two independent pipelines within ZenML. Each pipeline works on its own, creating specific artifacts. But what's fascinating is that these seemingly separate pipelines are intricately connected, all with the goal of delivering precise predictions.\n",
    "\n",
    "Before the Model Control Plane, connecting these pipelines and consolidating everything was a challenge. Imagine extracting a trained model artifact from the training pipeline and smoothly integrating it into the predictions pipeline. Previously, this involved complex ID references, leading to constant config updates, or blindly relying on the latest training run. But what if that run didn't meet the necessary performance standards? Using a subpar model for predictions was out of the question, especially for vital applications!\n",
    "\n",
    "Enter the Model Control Plane. This feature empowers you to effortlessly group pipelines, artifacts, and crucial business data into a unified entity: a `Model`. A Model captures lineage information and more. Within a Model, different `Model Versions` can be staged. For example, you can rely on your predictions at a specific stage, like `Production``, and decide whether the Model Version should be promoted based on your business rules during training. Plus, accessing data from other Models and their Versions is just as simple, enhancing the system's adaptability."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Overview of the process\n",
    "\n",
    "![Pipelines Overview](_assets/train_prediction_example.png)\n",
    "\n",
    "Each time the `train_and_promote` pipeline runs, it creates a new iris_classifier. However, it only promotes the created model to `Production` if a certain accuracy threshold is met. The `do_predictions` pipeline simply picks up the latest Promoted model and runs batch inference on it. That way these two pipelines can independently be run, but can rely on each others output."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Installation and Initialization\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from zenml.environment import Environment\n",
    "\n",
    "if Environment.in_google_colab():\n",
    "    # Install Cloudflare Tunnel binary\n",
    "    !wget -q https://github.com/cloudflare/cloudflared/releases/latest/download/cloudflared-linux-amd64.deb && dpkg -i cloudf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Warning: This will remove all your local ZenML artifacts. If you don't want this then just remove the zenml clean command"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: zenml[dev,server]>=0.45.5 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (0.45.5)\n",
      "Requirement already satisfied: alembic<1.9.0,>=1.8.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (1.8.1)\n",
      "Requirement already satisfied: azure-mgmt-resource>=21.0.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (23.0.1)\n",
      "Requirement already satisfied: click<8.1.4,>=8.0.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (8.1.3)\n",
      "Requirement already satisfied: click-params<0.4.0,>=0.3.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (0.3.0)\n",
      "Requirement already satisfied: cloudpickle<3,>=2.0.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (2.2.1)\n",
      "Requirement already satisfied: distro<2.0.0,>=1.6.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (1.8.0)\n",
      "Requirement already satisfied: docker<6.2.0,>=6.1.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (6.1.3)\n",
      "Requirement already satisfied: gitpython<4.0.0,>=3.1.18 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (3.1.40)\n",
      "Requirement already satisfied: httplib2<0.20,>=0.19.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (0.19.1)\n",
      "Requirement already satisfied: pandas>=1.1.5 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (2.0.3)\n",
      "Requirement already satisfied: passlib[bcrypt]<1.8.0,>=1.7.4 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (1.7.4)\n",
      "Requirement already satisfied: psutil>=5.0.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (5.9.6)\n",
      "Requirement already satisfied: pydantic<1.11,>=1.9.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (1.10.13)\n",
      "Requirement already satisfied: pymysql<1.1.0,>=1.0.2 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (1.0.3)\n",
      "Requirement already satisfied: pyparsing<3,>=2.4.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (2.4.7)\n",
      "Requirement already satisfied: python-dateutil<3.0.0,>=2.8.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (2.8.2)\n",
      "Requirement already satisfied: pyyaml>=6.0.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (6.0.1)\n",
      "Requirement already satisfied: rich[jupyter]>=12.0.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (13.6.0)\n",
      "Requirement already satisfied: sqlalchemy_utils==0.38.3 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (0.38.3)\n",
      "Requirement already satisfied: sqlmodel==0.0.8 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (0.0.8)\n",
      "Requirement already satisfied: Jinja2 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (3.1.2)\n",
      "Requirement already satisfied: fastapi<0.100,>=0.75 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (0.99.1)\n",
      "Requirement already satisfied: fastapi-utils<0.3.0,>=0.2.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (0.2.1)\n",
      "Requirement already satisfied: ipinfo>=4.4.3 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (4.4.3)\n",
      "Requirement already satisfied: orjson<3.9.0,>=3.8.3 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (3.8.14)\n",
      "Requirement already satisfied: pyjwt[crypto]==2.7.* in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (2.7.0)\n",
      "Requirement already satisfied: python-multipart<0.1.0,>=0.0.5 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (0.0.6)\n",
      "Requirement already satisfied: uvicorn[standard]>=0.17.5 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (0.23.2)\n",
      "Requirement already satisfied: bandit<2.0.0,>=1.7.5 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (1.7.5)\n",
      "Requirement already satisfied: black<24.0.0,>=23.10.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (23.10.1)\n",
      "Requirement already satisfied: coverage[toml]<6.0,>=5.5 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (5.5)\n",
      "Requirement already satisfied: darglint<2.0.0,>=1.8.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (1.8.1)\n",
      "Requirement already satisfied: hypothesis<7.0.0,>=6.43.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (6.88.1)\n",
      "Requirement already satisfied: mike<2.0.0,>=1.1.2 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (1.1.2)\n",
      "Requirement already satisfied: mkdocs<2.0.0,>=1.2.3 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (1.5.3)\n",
      "Requirement already satisfied: mkdocs-awesome-pages-plugin<3.0.0,>=2.6.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (2.9.2)\n",
      "Requirement already satisfied: mkdocs-material<9.0.0,>=8.1.7 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (8.5.11)\n",
      "Requirement already satisfied: mkdocstrings<0.18.0,>=0.17.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (0.17.0)\n",
      "Requirement already satisfied: mypy==1.6.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (1.6.1)\n",
      "Requirement already satisfied: pre-commit<3.0.0,>=2.14.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (2.21.0)\n",
      "Requirement already satisfied: pyment<0.4.0,>=0.3.3 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (0.3.3)\n",
      "Requirement already satisfied: pytest<8.0.0,>=7.4.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (7.4.3)\n",
      "Requirement already satisfied: pytest-clarity<2.0.0,>=1.0.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (1.0.1)\n",
      "Requirement already satisfied: pytest-mock<4.0.0,>=3.6.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (3.12.0)\n",
      "Requirement already satisfied: pytest-randomly<4.0.0,>=3.10.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (3.15.0)\n",
      "Requirement already satisfied: ruff<0.2.0,>=0.1.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (0.1.3)\n",
      "Requirement already satisfied: tox<4.0.0,>=3.24.3 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (3.28.0)\n",
      "Requirement already satisfied: types-Markdown<4.0.0,>=3.3.6 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (3.5.0.0)\n",
      "Requirement already satisfied: types-Pillow<10.0.0,>=9.2.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (9.5.0.6)\n",
      "Requirement already satisfied: types-PyMySQL<2.0.0,>=1.0.4 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (1.1.0.1)\n",
      "Requirement already satisfied: types-PyYAML<7.0.0,>=6.0.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (6.0.12.12)\n",
      "Requirement already satisfied: types-certifi<2022.0.0.0,>=2021.10.8.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (2021.10.8.3)\n",
      "Requirement already satisfied: types-croniter<2.0.0,>=1.0.2 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (1.4.0.1)\n",
      "Requirement already satisfied: types-futures<4.0.0,>=3.3.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (3.3.8)\n",
      "Requirement already satisfied: types-passlib<2.0.0,>=1.7.7 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (1.7.7.13)\n",
      "Requirement already satisfied: types-protobuf<4.0.0,>=3.18.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (3.20.4.6)\n",
      "Requirement already satisfied: types-psutil<6.0.0,>=5.8.13 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (5.9.5.17)\n",
      "Requirement already satisfied: types-python-dateutil<3.0.0,>=2.8.2 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (2.8.19.14)\n",
      "Requirement already satisfied: types-python-slugify<6.0.0,>=5.0.2 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (5.0.4)\n",
      "Requirement already satisfied: types-redis<5.0.0,>=4.1.19 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (4.6.0.8)\n",
      "Requirement already satisfied: types-requests<3.0.0,>=2.27.11 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (2.31.0.10)\n",
      "Requirement already satisfied: types-setuptools<58.0.0,>=57.4.2 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (57.4.18)\n",
      "Requirement already satisfied: types-six<2.0.0,>=1.16.2 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (1.16.21.9)\n",
      "Requirement already satisfied: types-termcolor<2.0.0,>=1.1.2 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (1.1.6.2)\n",
      "Requirement already satisfied: typing-extensions>=3.7.4 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from zenml[dev,server]>=0.45.5) (4.8.0)\n",
      "Requirement already satisfied: mypy-extensions>=1.0.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from mypy==1.6.1->zenml[dev,server]>=0.45.5) (1.0.0)\n",
      "Requirement already satisfied: tomli>=1.1.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from mypy==1.6.1->zenml[dev,server]>=0.45.5) (2.0.1)\n",
      "Requirement already satisfied: cryptography>=3.4.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from pyjwt[crypto]==2.7.*->zenml[dev,server]>=0.45.5) (41.0.5)\n",
      "Requirement already satisfied: SQLAlchemy>=1.3 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from sqlalchemy_utils==0.38.3->zenml[dev,server]>=0.45.5) (1.4.41)\n",
      "Requirement already satisfied: sqlalchemy2-stubs in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from sqlmodel==0.0.8->zenml[dev,server]>=0.45.5) (0.0.2a35)\n",
      "Requirement already satisfied: Mako in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from alembic<1.9.0,>=1.8.1->zenml[dev,server]>=0.45.5) (1.2.4)\n",
      "Requirement already satisfied: importlib-metadata in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from alembic<1.9.0,>=1.8.1->zenml[dev,server]>=0.45.5) (6.8.0)\n",
      "Requirement already satisfied: importlib-resources in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from alembic<1.9.0,>=1.8.1->zenml[dev,server]>=0.45.5) (6.1.0)\n",
      "Requirement already satisfied: isodate<1.0.0,>=0.6.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from azure-mgmt-resource>=21.0.0->zenml[dev,server]>=0.45.5) (0.6.1)\n",
      "Requirement already satisfied: azure-common~=1.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from azure-mgmt-resource>=21.0.0->zenml[dev,server]>=0.45.5) (1.1.28)\n",
      "Requirement already satisfied: azure-mgmt-core<2.0.0,>=1.3.2 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from azure-mgmt-resource>=21.0.0->zenml[dev,server]>=0.45.5) (1.4.0)\n",
      "Requirement already satisfied: stevedore>=1.20.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from bandit<2.0.0,>=1.7.5->zenml[dev,server]>=0.45.5) (5.1.0)\n",
      "Requirement already satisfied: packaging>=22.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from black<24.0.0,>=23.10.0->zenml[dev,server]>=0.45.5) (23.2)\n",
      "Requirement already satisfied: pathspec>=0.9.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from black<24.0.0,>=23.10.0->zenml[dev,server]>=0.45.5) (0.11.2)\n",
      "Requirement already satisfied: platformdirs>=2 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from black<24.0.0,>=23.10.0->zenml[dev,server]>=0.45.5) (3.11.0)\n",
      "Requirement already satisfied: validators<0.19,>=0.18 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from click-params<0.4.0,>=0.3.0->zenml[dev,server]>=0.45.5) (0.18.2)\n",
      "Requirement already satisfied: toml in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from coverage[toml]<6.0,>=5.5->zenml[dev,server]>=0.45.5) (0.10.2)\n",
      "Requirement already satisfied: requests>=2.26.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from docker<6.2.0,>=6.1.0->zenml[dev,server]>=0.45.5) (2.31.0)\n",
      "Requirement already satisfied: urllib3>=1.26.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from docker<6.2.0,>=6.1.0->zenml[dev,server]>=0.45.5) (2.0.7)\n",
      "Requirement already satisfied: websocket-client>=0.32.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from docker<6.2.0,>=6.1.0->zenml[dev,server]>=0.45.5) (1.6.4)\n",
      "Requirement already satisfied: starlette<0.28.0,>=0.27.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from fastapi<0.100,>=0.75->zenml[dev,server]>=0.45.5) (0.27.0)\n",
      "Requirement already satisfied: gitdb<5,>=4.0.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from gitpython<4.0.0,>=3.1.18->zenml[dev,server]>=0.45.5) (4.0.11)\n",
      "Requirement already satisfied: attrs>=19.2.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from hypothesis<7.0.0,>=6.43.1->zenml[dev,server]>=0.45.5) (23.1.0)\n",
      "Requirement already satisfied: sortedcontainers<3.0.0,>=2.1.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from hypothesis<7.0.0,>=6.43.1->zenml[dev,server]>=0.45.5) (2.4.0)\n",
      "Requirement already satisfied: exceptiongroup>=1.0.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from hypothesis<7.0.0,>=6.43.1->zenml[dev,server]>=0.45.5) (1.1.3)\n",
      "Requirement already satisfied: cachetools in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from ipinfo>=4.4.3->zenml[dev,server]>=0.45.5) (5.3.2)\n",
      "Requirement already satisfied: aiohttp<=4 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from ipinfo>=4.4.3->zenml[dev,server]>=0.45.5) (3.8.6)\n",
      "Requirement already satisfied: verspec in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from mike<2.0.0,>=1.1.2->zenml[dev,server]>=0.45.5) (0.1.0)\n",
      "Requirement already satisfied: ghp-import>=1.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from mkdocs<2.0.0,>=1.2.3->zenml[dev,server]>=0.45.5) (2.1.0)\n",
      "Requirement already satisfied: markdown>=3.2.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from mkdocs<2.0.0,>=1.2.3->zenml[dev,server]>=0.45.5) (3.5)\n",
      "Requirement already satisfied: markupsafe>=2.0.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from mkdocs<2.0.0,>=1.2.3->zenml[dev,server]>=0.45.5) (2.1.3)\n",
      "Requirement already satisfied: mergedeep>=1.3.4 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from mkdocs<2.0.0,>=1.2.3->zenml[dev,server]>=0.45.5) (1.3.4)\n",
      "Requirement already satisfied: pyyaml-env-tag>=0.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from mkdocs<2.0.0,>=1.2.3->zenml[dev,server]>=0.45.5) (0.1)\n",
      "Requirement already satisfied: watchdog>=2.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from mkdocs<2.0.0,>=1.2.3->zenml[dev,server]>=0.45.5) (3.0.0)\n",
      "Requirement already satisfied: natsort>=8.1.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from mkdocs-awesome-pages-plugin<3.0.0,>=2.6.1->zenml[dev,server]>=0.45.5) (8.4.0)\n",
      "Requirement already satisfied: wcmatch>=7 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from mkdocs-awesome-pages-plugin<3.0.0,>=2.6.1->zenml[dev,server]>=0.45.5) (8.5)\n",
      "Requirement already satisfied: mkdocs-material-extensions>=1.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from mkdocs-material<9.0.0,>=8.1.7->zenml[dev,server]>=0.45.5) (1.3)\n",
      "Requirement already satisfied: pygments>=2.12 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from mkdocs-material<9.0.0,>=8.1.7->zenml[dev,server]>=0.45.5) (2.16.1)\n",
      "Requirement already satisfied: pymdown-extensions>=9.4 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from mkdocs-material<9.0.0,>=8.1.7->zenml[dev,server]>=0.45.5) (10.3.1)\n",
      "Requirement already satisfied: mkdocs-autorefs>=0.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from mkdocstrings<0.18.0,>=0.17.0->zenml[dev,server]>=0.45.5) (0.5.0)\n",
      "Requirement already satisfied: pytkdocs>=0.14.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from mkdocstrings<0.18.0,>=0.17.0->zenml[dev,server]>=0.45.5) (0.16.1)\n",
      "Requirement already satisfied: pytz>=2020.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from pandas>=1.1.5->zenml[dev,server]>=0.45.5) (2023.3.post1)\n",
      "Requirement already satisfied: tzdata>=2022.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from pandas>=1.1.5->zenml[dev,server]>=0.45.5) (2023.3)\n",
      "Requirement already satisfied: numpy>=1.20.3 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from pandas>=1.1.5->zenml[dev,server]>=0.45.5) (1.24.4)\n",
      "Requirement already satisfied: bcrypt>=3.1.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from passlib[bcrypt]<1.8.0,>=1.7.4->zenml[dev,server]>=0.45.5) (4.0.1)\n",
      "Requirement already satisfied: cfgv>=2.0.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from pre-commit<3.0.0,>=2.14.0->zenml[dev,server]>=0.45.5) (3.4.0)\n",
      "Requirement already satisfied: identify>=1.0.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from pre-commit<3.0.0,>=2.14.0->zenml[dev,server]>=0.45.5) (2.5.30)\n",
      "Requirement already satisfied: nodeenv>=0.11.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from pre-commit<3.0.0,>=2.14.0->zenml[dev,server]>=0.45.5) (1.8.0)\n",
      "Requirement already satisfied: virtualenv>=20.10.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from pre-commit<3.0.0,>=2.14.0->zenml[dev,server]>=0.45.5) (20.24.6)\n",
      "Requirement already satisfied: iniconfig in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from pytest<8.0.0,>=7.4.0->zenml[dev,server]>=0.45.5) (2.0.0)\n",
      "Requirement already satisfied: pluggy<2.0,>=0.12 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from pytest<8.0.0,>=7.4.0->zenml[dev,server]>=0.45.5) (1.3.0)\n",
      "Requirement already satisfied: pprintpp>=0.4.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from pytest-clarity<2.0.0,>=1.0.1->zenml[dev,server]>=0.45.5) (0.4.0)\n",
      "Requirement already satisfied: six>=1.5 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from python-dateutil<3.0.0,>=2.8.1->zenml[dev,server]>=0.45.5) (1.16.0)\n",
      "Requirement already satisfied: markdown-it-py>=2.2.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from rich[jupyter]>=12.0.0->zenml[dev,server]>=0.45.5) (3.0.0)\n",
      "Requirement already satisfied: ipywidgets<9,>=7.5.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from rich[jupyter]>=12.0.0->zenml[dev,server]>=0.45.5) (8.1.1)\n",
      "Requirement already satisfied: filelock>=3.0.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from tox<4.0.0,>=3.24.3->zenml[dev,server]>=0.45.5) (3.12.4)\n",
      "Requirement already satisfied: py>=1.4.17 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from tox<4.0.0,>=3.24.3->zenml[dev,server]>=0.45.5) (1.11.0)\n",
      "Requirement already satisfied: types-pyOpenSSL in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from types-redis<5.0.0,>=4.1.19->zenml[dev,server]>=0.45.5) (23.2.0.2)\n",
      "Requirement already satisfied: h11>=0.8 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from uvicorn[standard]>=0.17.5->zenml[dev,server]>=0.45.5) (0.14.0)\n",
      "Requirement already satisfied: httptools>=0.5.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from uvicorn[standard]>=0.17.5->zenml[dev,server]>=0.45.5) (0.6.1)\n",
      "Requirement already satisfied: python-dotenv>=0.13 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from uvicorn[standard]>=0.17.5->zenml[dev,server]>=0.45.5) (1.0.0)\n",
      "Requirement already satisfied: uvloop!=0.15.0,!=0.15.1,>=0.14.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from uvicorn[standard]>=0.17.5->zenml[dev,server]>=0.45.5) (0.19.0)\n",
      "Requirement already satisfied: watchfiles>=0.13 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from uvicorn[standard]>=0.17.5->zenml[dev,server]>=0.45.5) (0.21.0)\n",
      "Requirement already satisfied: websockets>=10.4 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from uvicorn[standard]>=0.17.5->zenml[dev,server]>=0.45.5) (12.0)\n",
      "Requirement already satisfied: charset-normalizer<4.0,>=2.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from aiohttp<=4->ipinfo>=4.4.3->zenml[dev,server]>=0.45.5) (3.3.1)\n",
      "Requirement already satisfied: multidict<7.0,>=4.5 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from aiohttp<=4->ipinfo>=4.4.3->zenml[dev,server]>=0.45.5) (6.0.4)\n",
      "Requirement already satisfied: async-timeout<5.0,>=4.0.0a3 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from aiohttp<=4->ipinfo>=4.4.3->zenml[dev,server]>=0.45.5) (4.0.3)\n",
      "Requirement already satisfied: yarl<2.0,>=1.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from aiohttp<=4->ipinfo>=4.4.3->zenml[dev,server]>=0.45.5) (1.9.2)\n",
      "Requirement already satisfied: frozenlist>=1.1.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from aiohttp<=4->ipinfo>=4.4.3->zenml[dev,server]>=0.45.5) (1.4.0)\n",
      "Requirement already satisfied: aiosignal>=1.1.2 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from aiohttp<=4->ipinfo>=4.4.3->zenml[dev,server]>=0.45.5) (1.3.1)\n",
      "Requirement already satisfied: azure-core<2.0.0,>=1.26.2 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from azure-mgmt-core<2.0.0,>=1.3.2->azure-mgmt-resource>=21.0.0->zenml[dev,server]>=0.45.5) (1.29.5)\n",
      "Requirement already satisfied: cffi>=1.12 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from cryptography>=3.4.0->pyjwt[crypto]==2.7.*->zenml[dev,server]>=0.45.5) (1.16.0)\n",
      "Requirement already satisfied: smmap<6,>=3.0.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from gitdb<5,>=4.0.1->gitpython<4.0.0,>=3.1.18->zenml[dev,server]>=0.45.5) (5.0.1)\n",
      "Requirement already satisfied: zipp>=0.5 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from importlib-metadata->alembic<1.9.0,>=1.8.1->zenml[dev,server]>=0.45.5) (3.17.0)\n",
      "Requirement already satisfied: comm>=0.1.3 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from ipywidgets<9,>=7.5.1->rich[jupyter]>=12.0.0->zenml[dev,server]>=0.45.5) (0.1.4)\n",
      "Requirement already satisfied: ipython>=6.1.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from ipywidgets<9,>=7.5.1->rich[jupyter]>=12.0.0->zenml[dev,server]>=0.45.5) (8.12.3)\n",
      "Requirement already satisfied: traitlets>=4.3.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from ipywidgets<9,>=7.5.1->rich[jupyter]>=12.0.0->zenml[dev,server]>=0.45.5) (5.12.0)\n",
      "Requirement already satisfied: widgetsnbextension~=4.0.9 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from ipywidgets<9,>=7.5.1->rich[jupyter]>=12.0.0->zenml[dev,server]>=0.45.5) (4.0.9)\n",
      "Requirement already satisfied: jupyterlab-widgets~=3.0.9 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from ipywidgets<9,>=7.5.1->rich[jupyter]>=12.0.0->zenml[dev,server]>=0.45.5) (3.0.9)\n",
      "Requirement already satisfied: mdurl~=0.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from markdown-it-py>=2.2.0->rich[jupyter]>=12.0.0->zenml[dev,server]>=0.45.5) (0.1.2)\n",
      "Requirement already satisfied: setuptools in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from nodeenv>=0.11.1->pre-commit<3.0.0,>=2.14.0->zenml[dev,server]>=0.45.5) (68.2.2)\n",
      "Requirement already satisfied: astunparse>=1.6 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from pytkdocs>=0.14.0->mkdocstrings<0.18.0,>=0.17.0->zenml[dev,server]>=0.45.5) (1.6.3)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from requests>=2.26.0->docker<6.2.0,>=6.1.0->zenml[dev,server]>=0.45.5) (3.4)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from requests>=2.26.0->docker<6.2.0,>=6.1.0->zenml[dev,server]>=0.45.5) (2023.7.22)\n",
      "Requirement already satisfied: greenlet!=0.4.17 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from SQLAlchemy>=1.3->sqlalchemy_utils==0.38.3->zenml[dev,server]>=0.45.5) (3.0.1)\n",
      "Requirement already satisfied: anyio<5,>=3.4.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from starlette<0.28.0,>=0.27.0->fastapi<0.100,>=0.75->zenml[dev,server]>=0.45.5) (4.0.0)\n",
      "Requirement already satisfied: pbr!=2.1.0,>=2.0.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from stevedore>=1.20.0->bandit<2.0.0,>=1.7.5->zenml[dev,server]>=0.45.5) (5.11.1)\n",
      "Requirement already satisfied: decorator>=3.4.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from validators<0.19,>=0.18->click-params<0.4.0,>=0.3.0->zenml[dev,server]>=0.45.5) (5.1.1)\n",
      "Requirement already satisfied: distlib<1,>=0.3.7 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from virtualenv>=20.10.0->pre-commit<3.0.0,>=2.14.0->zenml[dev,server]>=0.45.5) (0.3.7)\n",
      "Requirement already satisfied: bracex>=2.1.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from wcmatch>=7->mkdocs-awesome-pages-plugin<3.0.0,>=2.6.1->zenml[dev,server]>=0.45.5) (2.4)\n",
      "Requirement already satisfied: sniffio>=1.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from anyio<5,>=3.4.0->starlette<0.28.0,>=0.27.0->fastapi<0.100,>=0.75->zenml[dev,server]>=0.45.5) (1.3.0)\n",
      "Requirement already satisfied: wheel<1.0,>=0.23.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from astunparse>=1.6->pytkdocs>=0.14.0->mkdocstrings<0.18.0,>=0.17.0->zenml[dev,server]>=0.45.5) (0.41.2)\n",
      "Requirement already satisfied: pycparser in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from cffi>=1.12->cryptography>=3.4.0->pyjwt[crypto]==2.7.*->zenml[dev,server]>=0.45.5) (2.21)\n",
      "Requirement already satisfied: backcall in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from ipython>=6.1.0->ipywidgets<9,>=7.5.1->rich[jupyter]>=12.0.0->zenml[dev,server]>=0.45.5) (0.2.0)\n",
      "Requirement already satisfied: jedi>=0.16 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from ipython>=6.1.0->ipywidgets<9,>=7.5.1->rich[jupyter]>=12.0.0->zenml[dev,server]>=0.45.5) (0.19.1)\n",
      "Requirement already satisfied: matplotlib-inline in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from ipython>=6.1.0->ipywidgets<9,>=7.5.1->rich[jupyter]>=12.0.0->zenml[dev,server]>=0.45.5) (0.1.6)\n",
      "Requirement already satisfied: pickleshare in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from ipython>=6.1.0->ipywidgets<9,>=7.5.1->rich[jupyter]>=12.0.0->zenml[dev,server]>=0.45.5) (0.7.5)\n",
      "Requirement already satisfied: prompt-toolkit!=3.0.37,<3.1.0,>=3.0.30 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from ipython>=6.1.0->ipywidgets<9,>=7.5.1->rich[jupyter]>=12.0.0->zenml[dev,server]>=0.45.5) (3.0.39)\n",
      "Requirement already satisfied: stack-data in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from ipython>=6.1.0->ipywidgets<9,>=7.5.1->rich[jupyter]>=12.0.0->zenml[dev,server]>=0.45.5) (0.6.3)\n",
      "Requirement already satisfied: pexpect>4.3 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from ipython>=6.1.0->ipywidgets<9,>=7.5.1->rich[jupyter]>=12.0.0->zenml[dev,server]>=0.45.5) (4.8.0)\n",
      "Requirement already satisfied: parso<0.9.0,>=0.8.3 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from jedi>=0.16->ipython>=6.1.0->ipywidgets<9,>=7.5.1->rich[jupyter]>=12.0.0->zenml[dev,server]>=0.45.5) (0.8.3)\n",
      "Requirement already satisfied: ptyprocess>=0.5 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from pexpect>4.3->ipython>=6.1.0->ipywidgets<9,>=7.5.1->rich[jupyter]>=12.0.0->zenml[dev,server]>=0.45.5) (0.7.0)\n",
      "Requirement already satisfied: wcwidth in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from prompt-toolkit!=3.0.37,<3.1.0,>=3.0.30->ipython>=6.1.0->ipywidgets<9,>=7.5.1->rich[jupyter]>=12.0.0->zenml[dev,server]>=0.45.5) (0.2.8)\n",
      "Requirement already satisfied: executing>=1.2.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from stack-data->ipython>=6.1.0->ipywidgets<9,>=7.5.1->rich[jupyter]>=12.0.0->zenml[dev,server]>=0.45.5) (2.0.0)\n",
      "Requirement already satisfied: asttokens>=2.1.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from stack-data->ipython>=6.1.0->ipywidgets<9,>=7.5.1->rich[jupyter]>=12.0.0->zenml[dev,server]>=0.45.5) (2.4.0)\n",
      "Requirement already satisfied: pure-eval in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from stack-data->ipython>=6.1.0->ipywidgets<9,>=7.5.1->rich[jupyter]>=12.0.0->zenml[dev,server]>=0.45.5) (0.2.2)\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.2.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.3.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n",
      "\u001b[2;36mThe local ZenML dashboard is not running.\u001b[0m\n",
      "\u001b[2;36mDeleted local ZenML config from \u001b[0m\n",
      "\u001b[2;35m/home/htahir1/workspace/zenml_io/zenml_playground/zenml-plugins/model_control_pl\u001b[0m\n",
      "\u001b[2;35mane/\u001b[0m\u001b[2;95m.zen.\u001b[0m\n",
      "\u001b[33mDeleting \u001b[0m\u001b[32m'local_stores'\u001b[0m\u001b[33m directory from global config.\u001b[0m\n",
      "\u001b[2;36mDeleted global ZenML config from \u001b[0m\u001b[2;35m/home/htahir1/.config/\u001b[0m\u001b[2;95mzenml.\u001b[0m\n",
      "\u001b[1;35mCreating default workspace 'default' ...\u001b[0m\n",
      "\u001b[33mThe current repo active workspace is no longer available. Resetting the active workspace to 'default'.\u001b[0m\n",
      "\u001b[33mThe current repo active stack is no longer available. Resetting the active stack to default.\u001b[0m\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/zenml/zen_stores/base_zen_store.py\", line 418, in validate_active_config\n",
      "    active_stack = self.get_stack(stack_id=active_stack_id)\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/zenml/zen_stores/sql_zen_store.py\", line 1100, in get_stack\n",
      "    raise KeyError(f\"Stack with ID {stack_id} not found.\")\n",
      "KeyError: 'Stack with ID 6ca4ec9c-60d1-4824-b4d3-563ffe658835 not found.'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/zenml/zen_stores/base_zen_store.py\", line 495, in _get_or_create_default_stack\n",
      "    user_name_or_id=self.get_user().id,\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/zenml/zen_stores/sql_zen_store.py\", line 1932, in get_user\n",
      "    user = self._get_user_schema(user_name_or_id, session=session)\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/zenml/zen_stores/sql_zen_store.py\", line 5403, in _get_user_schema\n",
      "    return self._get_schema_by_name_or_id(\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/zenml/zen_stores/sql_zen_store.py\", line 5359, in _get_schema_by_name_or_id\n",
      "    raise KeyError(error_msg)\n",
      "KeyError: \"Unable to get user with name or ID 'default': 'default' is not a valid  UUID and no user with this name exists.\"\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/zenml/zen_stores/base_zen_store.py\", line 138, in __init__\n",
      "    self._initialize()\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/zenml/zen_stores/sql_zen_store.py\", line 900, in _initialize\n",
      "    self.migrate_database()\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/zenml/zen_stores/sql_zen_store.py\", line 983, in migrate_database\n",
      "    self._sync_flavors()\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/zenml/zen_stores/sql_zen_store.py\", line 987, in _sync_flavors\n",
      "    FlavorRegistry().register_flavors(store=self)\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/zenml/stack/flavor_registry.py\", line 52, in register_flavors\n",
      "    self.register_builtin_flavors(store=store)\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/zenml/stack/flavor_registry.py\", line 113, in register_builtin_flavors\n",
      "    flavor_request_model = flavor().to_model(\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/zenml/stack/flavor.py\", line 155, in to_model\n",
      "    client = Client()\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/zenml/client.py\", line 278, in __call__\n",
      "    \"Client\", super().__call__(*args, **kwargs)\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/zenml/client.py\", line 325, in __init__\n",
      "    self._set_active_root(root)\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/zenml/client.py\", line 382, in _set_active_root\n",
      "    self._sanitize_config()\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/zenml/client.py\", line 405, in _sanitize_config\n",
      "    active_workspace, active_stack = self.zen_store.validate_active_config(\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/zenml/zen_stores/base_zen_store.py\", line 425, in validate_active_config\n",
      "    active_stack = self._get_or_create_default_stack(\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/zenml/zen_stores/base_zen_store.py\", line 500, in _get_or_create_default_stack\n",
      "    user_name_or_id=self.get_user().id,\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/zenml/zen_stores/sql_zen_store.py\", line 1932, in get_user\n",
      "    user = self._get_user_schema(user_name_or_id, session=session)\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/zenml/zen_stores/sql_zen_store.py\", line 5403, in _get_user_schema\n",
      "    return self._get_schema_by_name_or_id(\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/zenml/zen_stores/sql_zen_store.py\", line 5359, in _get_schema_by_name_or_id\n",
      "    raise KeyError(error_msg)\n",
      "KeyError: \"Unable to get user with name or ID 'default': 'default' is not a valid  UUID and no user with this name exists.\"\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/htahir1/.virtualenvs/demo/bin/zenml\", line 8, in <module>\n",
      "    sys.exit(cli())\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/click/core.py\", line 1130, in __call__\n",
      "    return self.main(*args, **kwargs)\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/click/core.py\", line 1055, in main\n",
      "    rv = self.invoke(ctx)\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/click/core.py\", line 1657, in invoke\n",
      "    return _process_result(sub_ctx.command.invoke(sub_ctx))\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/click/core.py\", line 1404, in invoke\n",
      "    return ctx.invoke(self.callback, **ctx.params)\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/click/core.py\", line 760, in invoke\n",
      "    return __callback(*args, **kwargs)\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/click/decorators.py\", line 26, in new_func\n",
      "    return f(get_current_context(), *args, **kwargs)\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/zenml/cli/base.py\", line 354, in clean\n",
      "    fresh_gc.set_default_store()\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/zenml/config/global_config.py\", line 590, in set_default_store\n",
      "    self._configure_store(default_store_cfg)\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/zenml/config/global_config.py\", line 381, in _configure_store\n",
      "    store = BaseZenStore.create_store(config, True, **kwargs)\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/zenml/zen_stores/base_zen_store.py\", line 254, in create_store\n",
      "    store = store_class(\n",
      "  File \"/home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages/zenml/zen_stores/base_zen_store.py\", line 159, in __init__\n",
      "    raise RuntimeError(\n",
      "RuntimeError: Error initializing sql store with URL 'sqlite:////home/htahir1/.config/zenml/local_stores/default_zen_store/zenml.db': \"Unable to get user with name or ID 'default': 'default' is not a valid  UUID and no user with this name exists.\"\n",
      "\u001b[2K\u001b[32m⠙\u001b[0m Installing integrations.....Requirement already satisfied: scikit-learn<1.3 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (1.2.2)\n",
      "Requirement already satisfied: numpy>=1.17.3 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from scikit-learn<1.3) (1.24.4)\n",
      "Requirement already satisfied: scipy>=1.3.2 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from scikit-learn<1.3) (1.10.1)\n",
      "Requirement already satisfied: joblib>=1.1.1 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from scikit-learn<1.3) (1.3.2)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /home/htahir1/.virtualenvs/demo/lib/python3.8/site-packages (from scikit-learn<1.3) (3.2.0)\n",
      "\u001b[2K\u001b[32m⠏\u001b[0m Installing integrations...\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.2.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.3.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n",
      "\u001b[2K\u001b[32m⠏\u001b[0m Installing integrations...\n",
      "\u001b[1A\u001b[2K"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'status': 'ok', 'restart': True}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "!pip install \"zenml[server,dev]>=0.45.5\"\n",
    "!zenml clean -y  # remove this if you dont want to lose local data\n",
    "!zenml integration install sklearn -y\n",
    "\n",
    "import IPython\n",
    "IPython.Application.instance().kernel.do_shutdown(restart=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[?25l\u001b[32m⠋\u001b[0m Initializing ZenML repository at \n",
      "/home/htahir1/workspace/zenml_io/zenml_playground/zenml-plugins/model_control_pl\n",
      "ane.\n",
      "\u001b[1;35mSetting the repo active workspace to 'default'.\u001b[0m\n",
      "\u001b[33mSetting the repo active stack to default.\u001b[0m\n",
      "\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[2;36mZenML repository initialized at \u001b[0m\n",
      "\u001b[2;35m/home/htahir1/workspace/zenml_io/zenml_playground/zenml-plugins/\u001b[0m\u001b[2;95mmodel_control_pl\u001b[0m\n",
      "\u001b[2;95mane.\u001b[0m\n",
      "\u001b[2;32m⠋\u001b[0m\u001b[2;36m Initializing ZenML repository at \u001b[0m\n",
      "\u001b[2;36m/home/htahir1/workspace/zenml_io/zenml_playground/zenml-plugins/model_control_pl\u001b[0m\n",
      "\u001b[2;36mane.\u001b[0m\n",
      "\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[32m⠋\u001b[0m Initializing ZenML repository at \n",
      "/home/htahir1/workspace/zenml_io/zenml_playground/zenml-plugins/model_control_pl\n",
      "ane.\n",
      "\n",
      "\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[2;36mThe local active stack was initialized to \u001b[0m\u001b[2;32m'default'\u001b[0m\u001b[2;36m. This local configuration \u001b[0m\n",
      "\u001b[2;36mwill only take effect when you're running ZenML from the initialized repository \u001b[0m\n",
      "\u001b[2;36mroot, or from a subdirectory. For more information on repositories and \u001b[0m\n",
      "\u001b[2;36mconfigurations, please visit \u001b[0m\n",
      "\u001b[2;4;94mhttps://docs.zenml.io/user-guide/starter-guide/understand-stacks.\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!rm -rf .zen\n",
    "!zenml init"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2;36mNo models found.\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!zenml model list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0.45.5'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from zenml import __version__\n",
    "\n",
    "__version__"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Training Pipeline\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1;35mCreation of new model version was requested, but no version name was explicitly provided. Setting \u001b[0m\u001b[1;36mversion\u001b[1;35m to \u001b[0m\u001b[1;36mrunning\u001b[1;35m.\u001b[0m\n",
      "\u001b[1;35mInitiating a new run for the pipeline: \u001b[0m\u001b[1;36mtrain_and_promote_model\u001b[1;35m.\u001b[0m\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'NoneType' object has no attribute 'to_model'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[18], line 24\u001b[0m\n\u001b[1;32m     21\u001b[0m     _, score \u001b[38;5;241m=\u001b[39m train_and_evaluate(train_data\u001b[38;5;241m=\u001b[39mtrain_data, test_data\u001b[38;5;241m=\u001b[39mtest_data)\n\u001b[1;32m     22\u001b[0m     promote_model(score\u001b[38;5;241m=\u001b[39mscore)\n\u001b[0;32m---> 24\u001b[0m \u001b[43mtrain_and_promote_model\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.virtualenvs/demo/lib/python3.8/site-packages/zenml/new/pipelines/pipeline.py:1497\u001b[0m, in \u001b[0;36mPipeline.__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1494\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mentrypoint(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m   1496\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mprepare(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[0;32m-> 1497\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_run\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_run_args\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.virtualenvs/demo/lib/python3.8/site-packages/zenml/new/pipelines/pipeline.py:597\u001b[0m, in \u001b[0;36mPipeline._run\u001b[0;34m(self, run_name, enable_cache, enable_artifact_metadata, enable_artifact_visualization, enable_step_logs, schedule, build, settings, step_configurations, extra, config_path, unlisted, prevent_build_reuse)\u001b[0m\n\u001b[1;32m    595\u001b[0m pipeline_id \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    596\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m register_pipeline:\n\u001b[0;32m--> 597\u001b[0m     pipeline_id \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_register\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpipeline_spec\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpipeline_spec\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39mid\n\u001b[1;32m    599\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    600\u001b[0m     logger\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPipeline \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mname\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m is unlisted.\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/.virtualenvs/demo/lib/python3.8/site-packages/zenml/new/pipelines/pipeline.py:1168\u001b[0m, in \u001b[0;36mPipeline._register\u001b[0;34m(self, pipeline_spec)\u001b[0m\n\u001b[1;32m   1163\u001b[0m version_hash \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_compute_unique_identifier(\n\u001b[1;32m   1164\u001b[0m     pipeline_spec\u001b[38;5;241m=\u001b[39mpipeline_spec\n\u001b[1;32m   1165\u001b[0m )\n\u001b[1;32m   1167\u001b[0m client \u001b[38;5;241m=\u001b[39m Client()\n\u001b[0;32m-> 1168\u001b[0m matching_pipelines \u001b[38;5;241m=\u001b[39m \u001b[43mclient\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mlist_pipelines\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1169\u001b[0m \u001b[43m    \u001b[49m\u001b[43mname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mname\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1170\u001b[0m \u001b[43m    \u001b[49m\u001b[43mversion_hash\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mversion_hash\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1171\u001b[0m \u001b[43m    \u001b[49m\u001b[43msize\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1172\u001b[0m \u001b[43m    \u001b[49m\u001b[43msort_by\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mdesc:created\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1173\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1174\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m matching_pipelines\u001b[38;5;241m.\u001b[39mtotal:\n\u001b[1;32m   1175\u001b[0m     registered_pipeline \u001b[38;5;241m=\u001b[39m matching_pipelines\u001b[38;5;241m.\u001b[39mitems[\u001b[38;5;241m0\u001b[39m]\n",
      "File \u001b[0;32m~/.virtualenvs/demo/lib/python3.8/site-packages/zenml/client.py:2605\u001b[0m, in \u001b[0;36mClient.list_pipelines\u001b[0;34m(self, sort_by, page, size, logical_operator, id, created, updated, name, version, version_hash, docstring, workspace_id, user_id)\u001b[0m\n\u001b[1;32m   2589\u001b[0m pipeline_filter_model \u001b[38;5;241m=\u001b[39m PipelineFilterModel(\n\u001b[1;32m   2590\u001b[0m     sort_by\u001b[38;5;241m=\u001b[39msort_by,\n\u001b[1;32m   2591\u001b[0m     page\u001b[38;5;241m=\u001b[39mpage,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   2602\u001b[0m     user_id\u001b[38;5;241m=\u001b[39muser_id,\n\u001b[1;32m   2603\u001b[0m )\n\u001b[1;32m   2604\u001b[0m pipeline_filter_model\u001b[38;5;241m.\u001b[39mset_scope_workspace(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mactive_workspace\u001b[38;5;241m.\u001b[39mid)\n\u001b[0;32m-> 2605\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mzen_store\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mlist_pipelines\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   2606\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpipeline_filter_model\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpipeline_filter_model\u001b[49m\n\u001b[1;32m   2607\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.virtualenvs/demo/lib/python3.8/site-packages/zenml/zen_stores/sql_zen_store.py:2863\u001b[0m, in \u001b[0;36mSqlZenStore.list_pipelines\u001b[0;34m(self, pipeline_filter_model)\u001b[0m\n\u001b[1;32m   2861\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m Session(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mengine) \u001b[38;5;28;01mas\u001b[39;00m session:\n\u001b[1;32m   2862\u001b[0m     query \u001b[38;5;241m=\u001b[39m select(PipelineSchema)\n\u001b[0;32m-> 2863\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfilter_and_paginate\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   2864\u001b[0m \u001b[43m        \u001b[49m\u001b[43msession\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msession\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2865\u001b[0m \u001b[43m        \u001b[49m\u001b[43mquery\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mquery\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2866\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtable\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mPipelineSchema\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2867\u001b[0m \u001b[43m        \u001b[49m\u001b[43mfilter_model\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpipeline_filter_model\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2868\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.virtualenvs/demo/lib/python3.8/site-packages/zenml/zen_stores/sql_zen_store.py:824\u001b[0m, in \u001b[0;36mSqlZenStore.filter_and_paginate\u001b[0;34m(cls, session, query, table, filter_model, custom_schema_to_model_conversion, custom_fetch)\u001b[0m\n\u001b[1;32m    822\u001b[0m to_model \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(schema, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mto_model\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m)\n\u001b[1;32m    823\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mcallable\u001b[39m(to_model):\n\u001b[0;32m--> 824\u001b[0m     items\u001b[38;5;241m.\u001b[39mappend(\u001b[43mto_model\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m)\n\u001b[1;32m    825\u001b[0m     \u001b[38;5;28;01mcontinue\u001b[39;00m\n\u001b[1;32m    826\u001b[0m \u001b[38;5;66;03m# If neither of the above work, raise an error.\u001b[39;00m\n",
      "File \u001b[0;32m~/.virtualenvs/demo/lib/python3.8/site-packages/zenml/zen_stores/schemas/pipeline_schemas.py:137\u001b[0m, in \u001b[0;36mPipelineSchema.to_model\u001b[0;34m(self, last_x_runs)\u001b[0m\n\u001b[1;32m    120\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mto_model\u001b[39m(\n\u001b[1;32m    121\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m    122\u001b[0m     last_x_runs: \u001b[38;5;28mint\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m3\u001b[39m,\n\u001b[1;32m    123\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPipelineResponseModel\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m    124\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Convert a `PipelineSchema` to a `PipelineModel`.\u001b[39;00m\n\u001b[1;32m    125\u001b[0m \n\u001b[1;32m    126\u001b[0m \u001b[38;5;124;03m    Args:\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    130\u001b[0m \u001b[38;5;124;03m        The created PipelineModel.\u001b[39;00m\n\u001b[1;32m    131\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[1;32m    132\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m PipelineResponseModel(\n\u001b[1;32m    133\u001b[0m         \u001b[38;5;28mid\u001b[39m\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mid,\n\u001b[1;32m    134\u001b[0m         name\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mname,\n\u001b[1;32m    135\u001b[0m         version\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mversion,\n\u001b[1;32m    136\u001b[0m         version_hash\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mversion_hash,\n\u001b[0;32m--> 137\u001b[0m         workspace\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mworkspace\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto_model\u001b[49m(),\n\u001b[1;32m    138\u001b[0m         user\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39muser\u001b[38;5;241m.\u001b[39mto_model(\u001b[38;5;28;01mTrue\u001b[39;00m) \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39muser \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m    139\u001b[0m         docstring\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdocstring,\n\u001b[1;32m    140\u001b[0m         spec\u001b[38;5;241m=\u001b[39mPipelineSpec\u001b[38;5;241m.\u001b[39mparse_raw(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mspec),\n\u001b[1;32m    141\u001b[0m         created\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcreated,\n\u001b[1;32m    142\u001b[0m         updated\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mupdated,\n\u001b[1;32m    143\u001b[0m         status\u001b[38;5;241m=\u001b[39m[run\u001b[38;5;241m.\u001b[39mstatus \u001b[38;5;28;01mfor\u001b[39;00m run \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mruns[:last_x_runs]],\n\u001b[1;32m    144\u001b[0m     )\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'NoneType' object has no attribute 'to_model'"
     ]
    }
   ],
   "source": [
    "from zenml import pipeline\n",
    "from zenml.model import ModelConfig\n",
    "\n",
    "from steps.train.load import load_data\n",
    "from steps.train.promote import promote_model\n",
    "from steps.train.train import train_and_evaluate\n",
    "\n",
    "\n",
    "@pipeline(\n",
    "    enable_cache=False,\n",
    "    model_config=ModelConfig(\n",
    "        name=\"iris_classifier\",\n",
    "        license=\"Apache\",\n",
    "        description=\"Show case Model Control Plane.\",\n",
    "        create_new_model_version=True,\n",
    "        delete_new_version_on_failure=True,\n",
    "    ),\n",
    ")\n",
    "def train_and_promote_model():\n",
    "    train_data, test_data = load_data()\n",
    "    _, score = train_and_evaluate(train_data=train_data, test_data=test_data)\n",
    "    promote_model(score=score)\n",
    "\n",
    "train_and_promote_model()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
